### PR TITLE
WIP: Simplify processors - add Fasttokenizers

### PR DESCRIPTION
**Pull request by Timoeller**
Simplifying the processor by:
- moving functions into dataset_from_dicts
- unnesting functions
- cleaning up code

------------------

Some older commits are done by Bogdan and me, for making FARM work with transformers 3.5.1 and fasttokenizers.

For descriptions about progress see comments below.

### CODE REVIEW

**Comment by Timoeller** [comment]
270996c introduces Multiprocessing after the Multithreading by Rust tokenizers. 

Although Multithreading should be finished by then, forking processes in python results in:
```
huggingface/tokenizers: The current process just got forked, after parallelism has already been used. Disabling parallelism to avoid deadlocks...
To disable this warning, you can either:
	- Avoid using `tokenizers` before the fork if possible
	- Explicitly set the environment variable TOKENIZERS_PARALLELISM=(true | false)
```

When setting the env var to false, python mp wont start.

My next experiment will be adding mp on a higher level (calling dataset_from_dicts) again.

**Comment by Timoeller** [comment]
I did some performance benchmarking and **found the culprit**:
The function offset_to_token_idx() in samples.py takes up 93% of compute time.
Vectorizing the function reduced preprocessing from 199 secs to 21 secs!

---------------------

I will now work on separating the create_samples_qa and sample_to_features_qa functions into more meaningful functions:
1. split_documents_into_passages
2. featurize_text_and_labels

**Comment by Timoeller** [comment]
Merged Brandens changes to the TextClassificationProcessor into this branch and also latest master.

**Comment by Timoeller** [comment]
Separated NQ from SQuAD, so we can focus on SQuAD refactoring only.
NQ only works with slow tokenizers for now.

**Comment by Timoeller** [comment]
QA processing tested by running question_answering_accuracy.py - using the Evaluator works as expected.

Working on inference mode now - this is still broken.

**Comment by Timoeller** [comment]
Inference does not really work as of now, it seems the answer is moved by 1 token to the right.

Tested #624 where we updated to transformers 3.5.1 and added fasttokenizers. There the inferencer works.

**Comment by Timoeller** [comment]
Found the bug! 
The bug was not only affecting inference but the whole QA processing pipeline:
We should not add special tokens when tokenizing documents and questions. These special tokens are added when qs and docs are combined.

**Comment by Timoeller** [comment]
Refactored sample to features for QA.

QA processing tested by running question_answering_accuracy.py - Evaluator and Inferencer work as expected.

File: farm/modeling/prediction_head.py [hunk 1]
```diff
@@ -690,41 +690,24 @@ def formatted_preds(self, logits, initial_mask, samples, return_class_probs=Fals
...
                 seq_res.append(
                     {
-                        "start": span["start"],
-                        "end": span["end"],
+                        "start": span[0],
+                        "end": span[1],
                         "context": f"{context}",
                         "label": f"{tag}",
-                        "probability": prob,
+                        "probability": np.float32(0.0),
```
Comment (Timoeller) [inline farm/modeling/prediction_head.py:710]: @brandenchan 
is the prob completly gone or was this just a quick fix that you forgot to revert?

**Comment by Timoeller** [comment]
I tested this branch with haystack in the following ways:

- running all haystack tests
- Tutorial 1 
- Tutorial 5, when using FarmReader.eval we need to adjust the return of processor.dataset_from_dicts
- Colab tutorial 5

Working on fixing the remaining tests.

**Comment by Timoeller** [comment]
I cannot reproduce the failing s3 test and nothing has changed code wise.

I presume it is some CI problem that we can fix later.

Merging now.

File: farm/modeling/prediction_head.py (hunk 1 above)
Comment (brandenchan) [reply farm/modeling/prediction_head.py:710]: The old probability calculation was wrong. I have opened issue #658 to address this.

**Pull request merged**